src/extractor/extractor.cpp

```diff
@@ -808,5 +818,173 @@ void Extractor::WriteCompressedNodeBasedGraph(const std::string &path,
+
+  while (i1 != v1.end() && i2 != v2.end())
+  {
+    if (i1->name_id == i2->name_id)
+    {
+      if (i1->name_id != EMPTY_NAMEID)
+        commons.push_back(std::make_pair(&(*i1), &(*i2)));
+
+      ++i1;
+      ++i2;
+    }
+    else if (i1->name_id < i2->name_id)
+      ++i1;
+    else
+      ++i2;
+  }
+
+  return (commons.size() >= 2);
+
+  /// @todo Process standalone U-turns.
+//  switch (commons.size())
+//  {
+//  case 0:
+//    return false;
+//  case 1:
+//    // ingoing + outgoing edges
+//    if (commons.front().first->direction + commons.front().second->direction != 1)
+//      return false;
+//  }
+
+//  return true;
+}
+
+void Extractor::FindSegregatedNodes(std::shared_ptr<util::NodeBasedDynamicGraph> graph,
+                                    std::vector<util::Coordinate> const & coordinates)
+{
+  auto const collectEdgeInfoFn = [&graph](auto const & edges1, NodeID node2)
+  {
+    std::vector<EdgeInfo> info;
+
+    for (auto const & e : edges1)
+    {
+      NodeID const target = graph->GetTarget(e);
+      if (target == node2)
+        continue;
+
+      auto const & data = graph->GetEdgeData(e);
+
+      BOOST_ASSERT(data.name_id != INVALID_NAMEID);
+      info.push_back({target, data.name_id, data.reversed ? 1 : 0, data.classes});
+    }
+
+    if (info.empty())
+      return info;
+
+    std::sort(info.begin(), info.end(), [](EdgeInfo const & e1, EdgeInfo const & e2)
+    {
+      return e1.node < e2.node;
+    });
+
+    // std::unique like code with custom combine routine.
+    auto curr = info.begin();
+    auto next = curr;
+    while (++next != info.end())
+    {
+      if (curr->node != next->node)
+      {
+        *(++curr) = *next;
+      }
+      else
+      {
+        BOOST_ASSERT(curr->name_id == next->name_id);
+        BOOST_ASSERT(curr->road_class == next->road_class);
+
+        BOOST_ASSERT(curr->direction != next->direction);
+        curr->direction = 2;
+      }
+    }
+
+    info.erase(++curr, info.end());
+
+    return info;
+  };
+
+  auto const isSegregatedFn = [&](auto const & edgeData,
+                                  auto const & edges1, NodeID node1,
+                                  auto const & edges2, NodeID node2)
+  {
+    return IsSegregated(collectEdgeInfoFn(edges1, node2), collectEdgeInfoFn(edges2, node1),
+                        {node1, edgeData.name_id, 0, edgeData.classes});
+  };
+
+  std::unordered_set<EdgeID> processed;
+
+  for (NodeID sourceID = 0; sourceID < graph->GetNumberOfNodes(); ++sourceID)
+  {
+    auto const sourceEdges = graph->GetAdjacentEdgeRange(sourceID);
+    for (EdgeID edgeID : sourceEdges)
+    {
+      auto & edgeData = graph->GetEdgeData(edgeID);
+
+      if (edgeData.reversed || edgeData.segregated || !processed.insert(edgeID).second)
+        continue;
+
+      NodeID const targetID = graph->GetTarget(edgeID);
+
+      if (osrm::util::coordinate_calculation::haversineDistance(coordinates[sourceID], coordinates[targetID]) > 30.0)
```

Comment:
this 30.0 is a heuristic from post-processing and will need to be revised. It is very restrictive but with many specific cases it just does not work, since road width can take up most of this 30 meters. See above that simply removing this check would not work.

Reply:
So you suggest remove this check, or ... ? I tried to run algorithm without this check and seems like it's ok, but sometimes we had situations like this:
```
a-b-b
|       |
|       |
a      b
|       |
|       |
a-b-b
```

a-a long way marked as segregated.

Reply:
No, I don't suggest getting rid of it completely. We need to ensure reasonable treatment, though. I would suggest considering the road-width of crossing roads here as well and to limit the length based on different scenarios.

E.g. if you have a segregated roat with 4 lanes per direction (not uncommon at intersections). You see 4 * 3.2 ~ 13 m per side. So (given that the ways are modelled in the middle of the road), we loose out on 13 meter for that threshold, limiting it to 20 meters. On small roads, we will allow larger differences. This is exactly opposite to what I would deem reasonable. Larger roads often have intermediate segments with traffic islands, while smaller roads have small traffic islands, if any. 
So this threshold should scale based on the local intersection and the involved roads, rather than be the fixed 30 meter from conservative post-processing.